File: src/lib/data-quality.ts
import { Site, SPV } from '@/types';

export type IssueSeverity = 'error' | 'warning' | 'info';

export interface DataQualityIssue {
  severity: IssueSeverity;
  code: string;
  siteId: string | null;
  siteName: string | null;
  message: string;
  fixable: boolean;
}

const ORDER: Record<IssueSeverity, number> = { error: 0, warning: 1, info: 2 };

function issue(
  severity: IssueSeverity,
  code: string,
  message: string,
  site?: Site,
  fixable = false
): DataQualityIssue {
  return {
    severity,
    code,
    siteId: site?.id ?? null,
    siteName: site?.name ?? null,
    message,
    fixable,
  };
}

export function checkSites(sites: Site[], spvs: SPV[], today = new Date()): DataQualityIssue[] {
  const known = Object.fromEntries(spvs.map((s) => [s.code.toUpperCase(), s]));
  const issues: DataQualityIssue[] = [];
  const names = new Map<string, Site[]>();

  for (const site of sites) {
    const key = (site.name || '').trim().toLowerCase();
    if (key) names.set(key, [...(names.get(key) || []), site]);
  }

  const todayIso = today.toISOString().slice(0, 10);

  for (const site of sites) {
    const size = site.systemSizeKwp || 0;
    const contracted = site.contractStatus === 'Yes';
    const code = site.spvCode;

    if (size <= 0) issues.push(issue('error', 'zero_size', 'System size is 0 kWp — fees cannot be calculated', site));

    if (code) {
      const upper = code.trim().toUpperCase();
      const spv = known[upper];
      if (!spv) {
        issues.push(issue('error', 'unknown_spv', `SPV code "${code}" does not match any configured SPV`, site));
      } else if (site.spvId !== spv.id || code !== upper) {
        issues.push(issue('warning', 'spv_link', `SPV link out of sync for code "${code}" (auto-fixable)`, site, true));
      }
    } else {
      issues.push(issue('warning', 'no_spv', 'No SPV assigned', site));
    }

    if (contracted && !site.onboardDate) {
      issues.push(issue('warning', 'no_onboard_date', 'Contracted site has no onboard date — excluded from CM Days tracking', site));
    }
    if (site.onboardDate && site.onboardDate.slice(0, 10) > todayIso) {
      issues.push(issue('warning', 'future_onboard', `Onboard date ${site.onboardDate.slice(0, 10)} is in the future`, site));
    }
    if (!contracted && site.onboardDate) {
      issues.push(issue('info', 'date_not_contracted', 'Has an onboard date but is not marked as contracted', site));
    }

    const costs = (site.pmCost || 0) + (site.cctvCost || 0) + (site.cleaningCost || 0);
    if (contracted && size > 0 && costs === 0) {
      issues.push(issue('info', 'zero_costs', 'Contracted site has £0 fixed costs (PM, CCTV, Cleaning)', site));
    }
    if (size > 100_000) {
      issues.push(issue('warning', 'large_size', `System size ${size.toLocaleString('en-GB')} kWp looks unusually large`, site));
    }
  }

  for (const group of names.values()) {
    if (group.length > 1) {
      for (const site of group) {
        issues.push(issue('warning', 'duplicate_name', `Site name appears ${group.length} times`, site));
      }
    }
  }

  return issues.sort((a, b) => ORDER[a.severity] - ORDER[b.severity] || (a.siteName || '').localeCompare(b.siteName || ''));
}

export function summariseIssues(issues: DataQualityIssue[]) {
  const counts = { error: 0, warning: 0, info: 0, total: issues.length, fixable: 0, sitesAffected: 0 };
  const sites = new Set<string>();
  for (const issue of issues) {
    counts[issue.severity] += 1;
    if (issue.fixable) counts.fixable += 1;
    if (issue.siteId) sites.add(issue.siteId);
  }
  counts.sitesAffected = sites.size;
  return counts;
}

File: legacy/streamlit/streamlit_app/data_quality.py
"""
Portfolio data-quality checks.

Runs a set of rules over the stored sites and returns a flat list of issues
that the dashboard, sites page and SPV page surface to the user.
"""

from __future__ import annotations

from datetime import date
from typing import Optional

import pandas as pd

SEVERITY_ORDER = {'error': 0, 'warning': 1, 'info': 2}


def _issue(severity: str, code: str, message: str, site: Optional[dict] = None, fixable: bool = False) -> dict:
    return {
        'severity': severity,
        'code': code,
        'site_id': site.get('id') if site else None,
        'site_name': site.get('name') if site else None,
        'message': message,
        'fixable': fixable,
    }


def check_sites(sites: list[dict], spvs: list[dict], today: Optional[date] = None) -> list[dict]:
    """Return every data-quality issue found, most severe first."""
    today = today or date.today()
    known_codes = {spv['code'].upper(): spv for spv in spvs}
    issues: list[dict] = []

    names: dict[str, list[dict]] = {}
    for site in sites:
        key = (site.get('name') or '').strip().lower()
        if key:
            names.setdefault(key, []).append(site)

    for site in sites:
        size = site.get('system_size_kwp') or 0
        contracted = site.get('contract_status') == 'Yes'
        code = site.get('spv_code')
        onboard = site.get('onboard_date')

        if size <= 0:
            issues.append(_issue('error', 'zero_size', 'System size is 0 kWp — fees cannot be calculated', site))

        if code:
            upper = code.strip().upper()
            spv = known_codes.get(upper)
            if not spv:
                issues.append(_issue(
                    'error', 'unknown_spv', f'SPV code "{code}" does not match any configured SPV', site
                ))
            elif site.get('spv_id') != spv['id'] or code != upper:
                issues.append(_issue(
                    'warning', 'spv_link', f'SPV link out of sync for code "{code}" (auto-fixable)', site, fixable=True
                ))
        else:
            issues.append(_issue('warning', 'no_spv', 'No SPV assigned', site))

        if contracted and not onboard:
            issues.append(_issue(
                'warning', 'no_onboard_date',
                'Contracted site has no onboard date — excluded from CM Days tracking', site,
            ))
        if onboard and onboard[:10] > today.isoformat():
            issues.append(_issue('warning', 'future_onboard', f'Onboard date {onboard[:10]} is in the future', site))
        if not contracted and onboard:
            issues.append(_issue('info', 'date_not_contracted', 'Has an onboard date but is not marked as contracted', site))

        costs = (site.get('pm_cost') or 0) + (site.get('cctv_cost') or 0) + (site.get('cleaning_cost') or 0)
        if contracted and size > 0 and costs == 0:
            issues.append(_issue('info', 'zero_costs', 'Contracted site has £0 fixed costs (PM, CCTV, Cleaning)', site))

        if size > 100_000:
            issues.append(_issue('warning', 'large_size', f'System size {size:,.0f} kWp looks unusually large', site))

    for group in names.values():
        if len(group) > 1:
            for site in group:
                issues.append(_issue(
                    'warning', 'duplicate_name', f'Site name appears {len(group)} times', site
                ))

    issues.sort(key=lambda i: (SEVERITY_ORDER.get(i['severity'], 9), i['site_name'] or ''))
    return issues


def summarise(issues: list[dict]) -> dict:
    counts = {'error': 0, 'warning': 0, 'info': 0}
    for issue in issues:
        counts[issue['severity']] = counts.get(issue['severity'], 0) + 1
    counts['total'] = len(issues)
    counts['fixable'] = sum(1 for i in issues if i['fixable'])
    counts['sites_affected'] = len({i['site_id'] for i in issues if i['site_id']})
    return counts


def issues_frame(issues: list[dict]) -> pd.DataFrame:
    if not issues:
        return pd.DataFrame(columns=['Severity', 'Site', 'Issue'])
    return pd.DataFrame([
        {
            'Severity': issue['severity'].title(),
            'Site': issue['site_name'] or '—',
            'Issue': issue['message'],
        }
        for issue in issues
    ])


def issues_for_site(issues: list[dict], site_id: str) -> list[dict]:
    return [i for i in issues if i['site_id'] == site_id]
